### PR TITLE
Added an Interface to allow for more nuanced IItemHandlers

### PR DESCRIPTION
**Pull request by Tschipp**
Adds #477 
Just a small interface which IItemHandlers of other mods should implement. 

It allows IItemHandlers to bypass the normal block counting methods and implement their own. This is useful for mods that have some special behavior which does not allow them to expose the exact contents of the IItemHandler, like for example an infinite supply of cobblestone. 

It was tested together with a (not yet public) version of Builder's Bag and was found to be working as expected.

### CODE REVIEW

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+public interface IBlockProviderItemHandler extends IItemHandler {
+
+	/**
+	 * Returns the amount of a certain ItemStack that this IItemHandler currently has
+	 * @param toCount The ItemStack to count
+	 * @param player The player
+	 * @return The count of this item in the IItemHandler
+	 */
+	public int getItemCount(ItemStack toCount, EntityPlayer player);
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:15]: I kind of don't like this Method - Reason in my comment to extractItems.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+	/**
+	 * Extracts the requested amount (or less) of the specified item from the IItemHandler.
+	 * @param toExtract The ItemStack to extract
+	 * @param maxCount The maximum count that should be extracted
+	 * @param player The player
+	 * @return The amount that was actually extracted
+	 */
+	public int extractItems(ItemStack toExtract, int maxCount, EntityPlayer player);
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:24]:
```suggestion
	public default int extractItems(EntityPlayer player, ItemStack toExtract, int maxCount, boolean simulate) {
	    int amountSaturated = 0;
	    for (int i = 0; i < getSlots(); i++) {

            ItemStack containerItem = inventory.getStackInSlot(i);

            if (containerItem.getItem() == target.getItem() && containerItem.getMetadata() == target.getMetadata()) {
                ItemStack stack = inventory.extractItem(i, amountRequired, simulate);
                amountSaturated += stack.getCount();

            }


            // Don't continue to check if we've saturated the amount.
            if( amountSaturated >= amountRequired )
                break;

        }
        
        return amountSaturated;
	}
```

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.direwolf20.buildinggadgets.common.integration;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.items.IItemHandler;
+
+public interface IBlockProviderItemHandler extends IItemHandler {
+
+	/**
+	 * Returns the amount of a certain ItemStack that this IItemHandler currently has
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:10]:
```suggestion
	 * Returns the amount of a certain {@link ItemStack} that this {@link IItemHandler} currently has without modifying it.
```

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 2 above)
Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:24]: Furthermore I'd kind of prefer it to have extractItems with a simulate boolean instead of the getItemCount Method. Especially as there might be inventory which have a different amount of Items, then can be extracted (locked slots for example).

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java [hunk 4]
```diff
@@ -130,11 +132,16 @@ public static boolean useItem(ItemStack target, EntityPlayer player, int amountR
         return amountLeft < amountRequired;
     }
 
-    private static int extractFromInventory(IItemHandler inventory, ItemStack target, int amountRequired) {
+    private static int extractFromInventory(IItemHandler inventory, ItemStack target, int amountRequired, EntityPlayer player) {
         int amountSaturated = 0;
         if( inventory == null )
             return amountSaturated;
 
+        if(inventory instanceof IBlockProviderItemHandler) {
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: Instead of short-curcuiting here for an IBlockProviderItemHandler, could you wrap non-IBlockProviderItemHandlers in a delegating adapter implementation and directly use it's Method? My suggestion for the default Method in the interface wouldn't require more then an implementation which delegates all of the regular IITemHandler Methods.

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java [hunk 5]
```diff
@@ -221,7 +230,12 @@ public static int countItem(ItemStack itemStack, EntityPlayer player, IRemoteInv
 
         if (invContainers.size() > 0) {
             for (IItemHandler container : invContainers) {
-                count += countInContainer(container, itemStack.getItem(), itemStack.getMetadata());
+            	{
+            		if(container instanceof IBlockProviderItemHandler)
+            			count += ((IBlockProviderItemHandler) container).getItemCount(itemStack, player);
+            		else
+            			count += countInContainer(container, itemStack.getItem(), itemStack.getMetadata());
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:237]: My suggestion would allow replacing this again with wrapping the container in an adapter and just passing simulate==false to the extractItems Method.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package com.direwolf20.buildinggadgets.common.integration;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.items.IItemHandler;
+
+public interface IBlockProviderItemHandler extends IItemHandler {
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:7]: I don't think this should be called `IBlockProviderItemHandler` as it has nothing to do with Blocks. It just adds an extended interface for accessing the items in the container. `IExtendedItemHandler`\`IAdvancedItemHandler` or `IItemAccess` maybe? The last is the best one of these 3 imo, as it lays the focus on it being used for accessing Items in BG.

Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:7]: Other suggestions?

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java (hunk 4 above)
Comment (Tschipp) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: I mean, sure. But I think it would increase code complexity by a lot and not make it more readable necessarily. Yes, it would make the code shorter overall, but I doubt it would make it more readable.

Comment (MichaelHillcox) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: My thoughts here extactly. I'd rather these special case inventories aren't dumped into the same logic as normal and typical inventories.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 3 above)
Comment (Tschipp) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:10]: I agree, the wording is a bit confusing. Additionally, we could say that it returns the amount which *can actually be removed*, to solve issues like Locked Slots.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 2 above)
Comment (Tschipp) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:24]: I'm not sure that getItemCount could be replaced by having an extra simulate boolean. I reckon that if simulate is true, we would still expect to get an exact count of how many items it should *simulate* to extract. However, in the place where getItemCount is called currently, we don't know about how many items we want to extract.
Unless of course, we augment the javadoc to clearly state that if simulate is true, no exact value must be provided which should be simulated. But I think that isn't really conventional and would cause unneccessary confusion.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 6 above)
Comment (Tschipp) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:7]: I agree, the name is a bit wordy. `IItemAccess` sounds good to me.

**Comment by MichaelHillcox** [comment]
Ay, thank you for making this :+1: I didn't think of doing it in such a simple way. I'll do a proper review in a few. Although it is true, I can't run 1.12 on my windows machine, I might still be able to use my Mac to compile and test so this shouldn't be an issue to merge in. Whilst I've got you, would you be able to hit me up on discord? My tag is: MiKeY#3704 I've got some related but not related questions :D If you've got time of course...

As for this, it seems fine although I want to check the return statement. The naming I agree with haha, Odd interface name but an easy fix.

**Comment by Tschipp** [comment]
Sure, I'm up for a chat! I sent you a friend request.

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java (hunk 4 above)
Comment (MichaelHillcox) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: You in this case, although I see the code advantages, I don't see a reason to go all over the top on a Limited Support version of the mod, as it is, I am only really wanting to work on 1.12 because of how much I like the Builder's bag mod :D and a few bugs I really want to fix.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 2 above)
Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:24]: getItemCount is currently used for the build preview so I'd say it should be as exact, as it can get.
Also weakening the simulate contract like this sounds reasonable to me, because you'd get the exact value anyway, once you extracted...

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java (hunk 4 above)
Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: I have no idea what you're trying to tell us here... sorry

Comment (MichaelHillcox) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java:140]: Essentially saying it's not worth the time nor effort to do that. I may modify things in the future but for now, na.

File: src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java (hunk 2 above)
Comment (MajorTuvok) [reply src/main/java/com/direwolf20/buildinggadgets/common/integration/IBlockProviderItemHandler.java:24]: What about instead renaming the `getItemCount` Method to something like `getItemsForExtraction` to signify what it shall report? That would be nicer then having a boolean argument telling the Method what it should return...
